sklearn/metrics/cluster/_expected_mutual_info_fast.pyx

```diff
@@ -35,36 +35,32 @@ def expected_mutual_information(contingency, int n_samples):
     nijs = np.arange(0, max(np.max(a), np.max(b)) + 1, dtype='float')
     nijs[0] = 1  # Stops divide by zero warnings. As its not used, no issue.
     # term1 is nij / N
-    term1 = nijs / N
+    term1 = nijs.base / n_samples
     # term2 is log((N*nij) / (a * b)) == log(N * nij) - log(a * b)
     log_a = np.log(a)
     log_b = np.log(b)
     # term2 uses log(N * nij) = log(N) + log(nij)
-    log_Nnij = np.log(N) + np.log(nijs)
+    log_Nnij = np.log(n_samples) + np.log(nijs)
     # term3 is large, and involved many factorials. Calculate these in log
     # space to stop overflows.
-    gln_a = gammaln(a + 1)
-    gln_b = gammaln(b + 1)
-    gln_Na = gammaln(N - a + 1)
-    gln_Nb = gammaln(N - b + 1)
-    gln_N = gammaln(N + 1)
-    gln_nij = gammaln(nijs + 1)
-    # start and end values for nij terms for each summation.
-    start = np.array([[v - N + w for w in b] for v in a], dtype='int')
-    start = np.maximum(start, 1)
-    end = np.minimum(np.resize(a, (C, R)).T, np.resize(b, (R, C))) + 1
+    gln_a = gammaln(a.base + 1)
+    gln_b = gammaln(b.base + 1)
+    gln_Na = gammaln(n_samples - a.base + 1)
+    gln_Nb = gammaln(n_samples - b.base + 1)
+    gln_Nnij = gammaln(nijs.base + 1) + gammaln(n_samples + 1)
```

Comment:
Could you please add an inline comment to explain how and why you pass the `.base` attribute in those lines? I must confess I do not understand what this does.

Reply:
I had to look up what `.base` is as well. https://numpy.org/doc/stable/reference/generated/numpy.ndarray.base.html says it is the base object, but even after reading that it didn't help me understand why this makes the code faster. Especially because I think for something like `x =np.arange(foo, bar)` I think `x.base` doesn't point to anything?

Reply:
This is not what makes the code faster. It's needed because `a`, `b` are now memoryviews instead of ndarrays. You can't do `a + 1` when `a` is a memoryview

Reply:
Maybe we need two comments then :D Your short comment explaining `.base` and one that briefly explains which part of this function to not change because it is the part that makes it fast?

Reply:
Is it really necessary to add a comment explaining `.base` specifically in this function while we're using it in many places of the code base ?